### PR TITLE
Migrate PermissionController and SubjectMetadataController

### PR DESCRIPTION
**Pull request by FrederikBolding**
Migrated `PermissionsController` and `SubjectMetadataController` from `snaps-skunkworks`.

**Description**

- ADDED:
  - Added `PermissionsController` and `SubjectMetadataController`
     - Several changes were needed to make it compatible with this repository. Added more doc strings, changed a few types in testing files. Also had to revert a recent testing change.
     - Added utility functions and types used by the controllers
 
Would probably recommend looking through some of the history when reviewing this. The very first commit is a 1:1 copy paste of the files from `snaps-skunkworks`. The following commits are the changes made to satisfy linting, making it build and have tests run successfully.

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

**Issue**

Resolves: https://github.com/MetaMask/snaps-skunkworks/issues/197

### CODE REVIEW

**Review by rekmarks — CHANGES_REQUESTED** [review]
Nice work! I can see why this took longer than expected. Just a handful of things to clear up.

File: src/util.ts [hunk 1]
```diff
@@ -944,10 +950,37 @@ export function isValidJson(value: unknown): value is Json {
...
+ * is called for purposes of POLA.
+ */
+export type HookNames<T> = {
+  [Property in keyof T]: true;
+};
+
+export type PermittedHandlerExport<T, U, V> = {
+  implementation: HandlerMiddlewareFunction<T, U, V>;
+  hookNames: HookNames<T>;
+} & BaseHandlerExport;
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/util.ts:984]: I think that might be problematic, as we might want to reference `@metamask/controllers` types in the RPC methods at some point. For example, when connecting the Snap RPC method to the new notification controller 😄

File: src/permissions/PermissionController.ts [hunk 2]
```diff
@@ -2075,7 +2074,7 @@ export class PermissionController<
    * Executes a restricted method as the subject with the given origin.
    * The specified params, if any, will be passed to the method implementation.
    *
-   * **ATTN:** Great caution should be exercised in the use of this method.
+   * ATTN:** Great caution should be exercised in the use of this method.
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/permissions/PermissionController.ts:2080]: Will see if it is possible to bold, otherwise remove.

File: src/permissions/Caveat.ts [hunk 3]
```diff
@@ -6,10 +6,9 @@ import {
   PermissionConstraint,
   RestrictedMethodParameters,
 } from './Permission';
-// This is used in a docstring, but ESLint doesn't notice it.
-/* eslint-disable @typescript-eslint/no-unused-vars */
+// @ts-expect-error This is used in a docstring
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/permissions/Caveat.ts:9]: This error specifically is because we reference the type in a JSDoc, we had lines to ignore the warning previously as well, I just changed it to fit with the rules of this repo (required a `ts-expect-error` instead).

The messenging system errors are a bit weird, but it seems to me that they are occurring because we are adding extra allowed actions to the messenger. I.e. access to the controller's own actions etc. So either we have a type error when passing it to the controller or when trying to use it for tests. I can dig up and post the errors in a bit.

File: src/permissions/PermissionController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/permissions/PermissionController.ts:2080]: Removed!

File: src/permissions/Caveat.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/permissions/Caveat.ts:9]: The messenging system type issues have been fixed in the latest commit

File: src/util.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/util.ts:984]: We already have a circular dependency between `@metamask/rpc-methods` and the permission controller in `@metamask/snap-controllers`, so this wouldn't be a new problem at least. But it doesn't seem ideal. Maybe better to move this type from `@metamask/rpc-methods` to the `@metamask/types` package, and reference it in both places?

File: src/permissions/Caveat.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/permissions/Caveat.ts:9]: Maybe worth disabling `noUnusedLocals` and `noUnusedParameters` in `tsconfig.json` in a separate PR. Those options predate us using ESLint in this repo. They are redundant alongside our ESLint rules.

File: src/permissions/PermissionController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/permissions/PermissionController.ts:2080]: I see that the trailing `**` was removed in at least one place, but they remain present on 4 lines (found 4 matches for `ATTN:**`)

File: src/permissions/rpc-methods/getPermissions.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+/**
+ * Get Permissions implementation to be used in JsonRpcEngine middleware.
+ *
+ * @param _req - The JsonRpcEngine request - unused
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options0 - Method hooks passed to the method implementation
```
Comment (Gudahtt) [inline src/permissions/rpc-methods/getPermissions.ts:35]: For some reason, the JSDoc ESLint plugin we use chooses `options0` as its default "object" name for destructured parameters. It's not a great name. I'd recommend changing it to `options` at least, or some better alternative.

There's another example of this in `requestPermissions.ts` as well.

File: src/util.ts [hunk 5]
```diff
@@ -882,3 +890,95 @@ export function getFormattedIpfsUrl(
...
+
+type PlainObject = Record<number | string | symbol, unknown>;
+
+/**
+ * Determines whether a value is a "plain" object.
+ *
+ * @param value - A value to check
+ * @returns True if the passed value is a plain object
+ */
+export function isPlainObject(value: unknown): value is PlainObject {
```
Comment (Gudahtt) [inline src/util.ts:902]: Could you add unit tests for these new utility functions? They are all showing as covered by the permission controller unit tests, but they should be tested directly as well to ensure all uses cases/boundary conditions are considered.

File: src/permissions/PermissionController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/permissions/PermissionController.ts:2080]: Removed!

File: src/permissions/rpc-methods/getPermissions.ts (hunk 4 above)
Comment (FrederikBolding) [reply src/permissions/rpc-methods/getPermissions.ts]: Fixed!

File: src/permissions/Caveat.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/permissions/Caveat.ts:9]: Yeah, I agree. But out of scope

File: src/util.ts (hunk 5 above)
Comment (FrederikBolding) [reply src/util.ts:902]: Done!

File: src/util.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/util.ts:984]: https://github.com/MetaMask/types/pull/13

File: src/permissions/Caveat.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rekmarks) [reply src/permissions/Caveat.ts:9]: #694

File: src/util.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rekmarks) [reply src/util.ts:984]: This PR should be good to go once we import those types `@metamask/types`.

**Review by rekmarks — APPROVED** [review]
LGTM!
